File: unixfs/file-info.go
package unixfs

import (
	"io/fs"
	"time"
)

// FileInfo contains information about a file.
type FileInfo struct {
	name    string
	size    int64
	mode    fs.FileMode
	modTime time.Time
}

// NewFileInfo constructs a new file info with details.
func NewFileInfo(
	name string,
	size int64,
	mode fs.FileMode,
	modTime time.Time,
) *FileInfo {
	return &FileInfo{
		name:    name,
		size:    size,
		mode:    mode,
		modTime: modTime,
	}
}

// Name returns the name of the file.
func (i *FileInfo) Name() string {
	return i.name
}

// Size returns length in bytes for regular files; system-dependent for others
func (i *FileInfo) Size() int64 {
	return i.size
}

// Mode is the unixfs file mode bitset.
func (i *FileInfo) Mode() fs.FileMode {
	return i.mode
}

// ModTime is the modification time.
func (i *FileInfo) ModTime() time.Time {
	return i.modTime
}

// IsDir is an abbreviation for Mode().IsDir()
func (i *FileInfo) IsDir() bool {
	return i.mode.IsDir()
}

// Sys returns the underlying data source (nil here).
func (i *FileInfo) Sys() any {
	return nil
}

// _ is a type assertion
var _ fs.FileInfo = ((*FileInfo)(nil))

File: db/unixfs/file-info.ts
// ModeDir is the directory bit in a file mode (matching Go os.ModeDir).
const ModeDir = 0x80000000

// FileInfo contains information about a file.
export class FileInfo {
  private readonly name: string
  private readonly size: bigint
  private readonly mode: number
  private readonly modTime: Date

  constructor(name: string, size: bigint, mode: number, modTime: Date) {
    this.name = name
    this.size = size
    this.mode = mode
    this.modTime = modTime
  }

  // getName returns the name of the file.
  getName(): string {
    return this.name
  }

  // getSize returns the length in bytes for regular files.
  getSize(): bigint {
    return this.size
  }

  // getMode returns the unixfs file mode bitset.
  getMode(): number {
    return this.mode
  }

  // getModTime returns the modification time.
  getModTime(): Date {
    return this.modTime
  }

  // isDir returns true if the mode indicates a directory.
  isDir(): boolean {
    return (this.mode & ModeDir) !== 0
  }
}
